Incompatible with using Revise

using ProtoStucts when Revise is loaded errors on method deletion.

Can't reproduce yet. More info needed.

I'm not sure if I have the same error: here a way to reproduce it (Julia 1.10 on Apple Silicon).

Create a file `"src/ProtoTest.jl"`:
```
using ProtoStructs

@proto struct Foo
    x :: Int
    # y :: Int
end
```
In a fresh Julia session:
```
julia> using Revise; includet("src/ProtoTest.jl")
```
Now uncomment the "y-line" in `"src/ProtoTest.jl"` and save it.

```
julia> Foo(42, 43)
┌ Error: Failed to revise /Users/.../ProtoTest.jl
│   exception =
│    AssertionError: ld[idx] < typemax(eltype(ld))
│    Stacktrace:
```

> julia> Foo(42, 43)
> ┌ Error: Failed to revise /Users/.../ProtoTest.jl
> │   exception =
> │    AssertionError: ld[idx] < typemax(eltype(ld))
> │    Stacktrace:
> │     [1] (::Revise.var"#76#78"{OrderedCollections.OrderedDict{Revise.RelocatableExpr, Union{Nothing, Vector{Any}}}, OrderedCollections.OrderedDict{Revise.RelocatableExpr, Union{Nothing, Vector{Any}}}})()
> │       @ Revise ~/.julia/packages/Revise/FaTes/src/packagedef.jl:269
> └ @ Revise ~/.julia/packages/Revise/FaTes/src/packagedef.jl:724

@TimHoly excuse me to ping you like this, but do you happen to know which assumption I am violating here?

--- a/src/ProtoStruct.jl
+++ b/src/ProtoStruct.jl
@@ -1,5 +1,29 @@
 
+const revise_uuid = Base.UUID("295af30f-e4ad-537b-8983-00126c2a3abe")
+const revise_pkgid = Base.PkgId(revise_uuid, "Revise")
+
+function checkrev()
+    revise_pkgid in keys(Base.loaded_modules) || return false 
+    d = @__DIR__ 
+    f = splitpath(@__FILE__)[end]
+
+    watched_files = getproperty(Base.loaded_modules[revise_pkgid], :watched_files)
+    d in keys(watched_files) || return false
+
+    trackedfiles = watched_files[d].trackedfiles
+    return f in keys(trackedfiles)
+end
+
 macro proto(expr)
+    if checkrev()
+        @eval __module__ $(_proto(expr))
+        return
+    else
+        return esc(_proto(expr))
+    end
+end
+
+function _proto(expr)
     if expr.head == :macrocall && expr.args[1] == Symbol("@kwdef")
         expr = expr.args[3]
     end
@@ -230,6 +254,6 @@ macro proto(expr)
                 end
             end
         end
-    return esc(ex)
+    return ex
 end
 

--- a/src/ProtoStruct.jl
+++ b/src/ProtoStruct.jl
@@ -6,10 +6,8 @@ function checkrev()
     revise_pkgid in keys(Base.loaded_modules) || return false 
     d = @__DIR__ 
     f = splitpath(@__FILE__)[end]
-
     watched_files = getproperty(Base.loaded_modules[revise_pkgid], :watched_files)
     d in keys(watched_files) || return false
-
     trackedfiles = watched_files[d].trackedfiles
     return f in keys(trackedfiles)
 end

